### PR TITLE
Sparse segment sum sqrtn op 

### PR DESCRIPTION
**Pull request by codeislife99**
This PR is for adding support for sparse segment sum sqrtn OP (https://www.tensorflow.org/api_docs/python/tf/sparse/segment_sum?hl=bn) as a part of a larger effort to add sparse operator support. (#7125, #7126)
This operation is TF Specific and can't be implemented using existing ops.

### CODE REVIEW

**Comment by codeislife99** [comment]
cc: @trevor-m @zhiics @comaniac @anijain2305 PTAL !

**Comment by comaniac** [comment]
The Relay part LGTM. However, since I'm not familiar with the implementation of those operators, I would ask @tkonolige and @mbrookhart to review this PR.

File: src/relay/op/tensor/transform.cc [hunk 1]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
     .set_attr<FTVMCompute>("FTVMCompute", MeshgridCompute)
     .set_attr<TOpPattern>("TOpPattern", kInjective);
 
+TVM_REGISTER_NODE_TYPE(SparseSegmentSumAttrs);
+
+bool SparseSegmentSumRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
+                         const TypeReporter& reporter) {
+  // types: [data, indices, segment_ids, result]
+  ICHECK_EQ(types.size(), 4) << "SparseSegmentSumRel expects 4 types but provided " << types.size();
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1561]:
```suggestion
  ICHECK_EQ(types.size(), 4) << "SparseSegmentSumRel expects 4 types but " << types.size() << " were provided.";
```

File: src/relay/op/tensor/transform.cc [hunk 2]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
...
+  auto data = types[0].as<TensorTypeNode>();
+  auto indices = types[1].as<TensorTypeNode>();
+  const auto* param = attrs.as<SparseSegmentSumAttrs>();
+  ICHECK(param != nullptr);
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1565]:
```suggestion
  ICHECK_NOTNULL(param);
```

File: src/relay/op/tensor/transform.cc [hunk 3]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
...
+  reporter->Assign(types[3], TupleType(Array<Type>(fields)));
+  return true;
+}
+
+Array<te::Tensor> SparseSegmentSumCompute(const Attrs& attrs, const Array<te::Tensor>& inputs,
+                                          const Type& out_type) {
+  ICHECK_EQ(inputs.size(), 3) << "SparseSegmentSumCompute expects 3 input but provided "
+                              << inputs.size();
+  const auto* param = attrs.as<SparseSegmentSumAttrs>();
+  ICHECK(param != nullptr);
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1583]:
```suggestion
  ICHECK_NOTNULL(param);
```

File: src/relay/op/tensor/transform.cc [hunk 4]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
...
+  fields.push_back(TensorType(new_data_shape, data->dtype));
+  fields.push_back(TensorType(Array<PrimExpr>{1}, tvm::DataType::Int(32)));
+  reporter->Assign(types[3], TupleType(Array<Type>(fields)));
+  return true;
+}
+
+Array<te::Tensor> SparseSegmentSumCompute(const Attrs& attrs, const Array<te::Tensor>& inputs,
+                                          const Type& out_type) {
+  ICHECK_EQ(inputs.size(), 3) << "SparseSegmentSumCompute expects 3 input but provided "
+                              << inputs.size();
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1581]:
```suggestion
  ICHECK_EQ(inputs.size(), 3) << "SparseSegmentSumCompute expects 3 input but " << inputs.size() << " were provided.";
```

File: src/relay/op/tensor/transform.cc [hunk 5]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
...
+TVM_REGISTER_GLOBAL("relay.op._make.sparse_segment_sum").set_body_typed(MakeSparseSegmentSum);
+
+RELAY_REGISTER_OP("sparse_segment_sum")
+    .describe(R"code(Return sparse segment sum of the tensor given segments
+)code" TVM_ADD_FILELINE)
+    .set_num_inputs(3)
+    .set_attrs_type<SparseSegmentSumAttrs>()
+    .add_argument("data", "Tensor", "The first tensor")
+    .add_argument("indices", "Tensor", "The second tensor")
+    .add_argument("segment_ids", "Tensor", "The third tensor")
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1603]: Please make the description better here.

File: src/relay/op/tensor/transform.cc [hunk 6]
```diff
@@ -1553,6 +1553,59 @@ RELAY_REGISTER_OP("meshgrid")
...
+    .add_type_rel("sparse_segment_sum", SparseSegmentSumRel)
+    .set_attr<TOpPattern>("TOpPattern", kInjective)
```
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1605]: @mbrookhart Is `kInjective` the correct pattern here?

File: python/tvm/relay/op/transform.py [hunk 7]
```diff
@@ -1320,3 +1320,55 @@ def adv_index(inputs):
         Output tensor.
     """
     return _make.adv_index(Tuple(inputs))
+
+
+def sparse_segment_sum(data, indices, segment_ids, num_segments=None):
+    """
+    Compute the sparse segment sum on the indices over the segment_ids
```
Comment (tkonolige) [inline python/tvm/relay/op/transform.py:1327]: I think we need a lot more documentation here. Please include a description of what a segment sum is and how a sparse segment sum is different from a dense one.

**Review by tkonolige — CHANGES_REQUESTED** [review]
One thing I missed on previous PRs, is that this is a new op. That means we should follow https://tvm.apache.org/docs/contribute/code_review.html#deliberate-on-api-and-data-structures. @codeislife99 could you do the following:
1. write up the differences between tf, pytorch, and mxnets's versions of this operation.
2. If this operation is TF specific, can it be implemented just using existing relay operators?
3. Answer these same questions for all the other sparse PRs you currently have open.

Given we are adding so many sparse operations, I think there are a couple of questions to answer:
1. Should we create a new module for sparse operations in `tvm.relay`? i.e. `tvm.relay.sparse`
2. Instead of hard coding add these specific one-off sparse operations, should we take a high level approach like XLA?

@tqchen @jroesch @ANSHUMAN87 (please ping anyone else who has been working on sparse, not sure if I got everyone).

It may be worth discussing this on discuss.

**Comment by ANSHUMAN87** [comment]
> One thing I missed on previous PRs, is that this is a new op. That means we should follow https://tvm.apache.org/docs/contribute/code_review.html#deliberate-on-api-and-data-structures. @codeislife99 could you do the following:
> 
>     1. write up the differences between tf, pytorch, and mxnets's versions of this operation.
> 
>     2. If this operation is TF specific, can it be implemented just using existing relay operators?
> 
>     3. Answer these same questions for all the other sparse PRs you currently have open.
> 
> 
> Given we are adding so many sparse operations, I think there are a couple of questions to answer:
> 
>     1. Should we create a new module for sparse operations in `tvm.relay`? i.e. `tvm.relay.sparse`
> 
>     2. Instead of hard coding add these specific one-off sparse operations, should we take a high level approach like XLA?
> 
> 
> @tqchen @jroesch @ANSHUMAN87 (please ping anyone else who has been working on sparse, not sure if I got everyone).
> 
> It may be worth discussing this on discuss.

Thanks @tkonolige! I am in total agreement with the points you have shared. All these points need to be concluded. 
I am already looking into these points as part of my plan for Sparse feature in TVM. Hopefully will be able to share a brief plan soon :slightly_smiling_face:

**Comment by antinucleon** [comment]
> One thing I missed on previous PRs, is that this is a new op. That means we should follow https://tvm.apache.org/docs/contribute/code_review.html#deliberate-on-api-and-data-structures. @codeislife99 could you do the following:
> 
> 1. write up the differences between tf, pytorch, and mxnets's versions of this operation.
> 2. If this operation is TF specific, can it be implemented just using existing relay operators?
> 3. Answer these same questions for all the other sparse PRs you currently have open.
> 
> Given we are adding so many sparse operations, I think there are a couple of questions to answer:
> 
> 1. Should we create a new module for sparse operations in `tvm.relay`? i.e. `tvm.relay.sparse`
> 2. Instead of hard coding add these specific one-off sparse operations, should we take a high level approach like XLA?
> 
> @tqchen @jroesch @ANSHUMAN87 (please ping anyone else who has been working on sparse, not sure if I got everyone).
> 
> It may be worth discussing this on discuss.

1. I am against to over document here. If the target is for TF frontend, why requires comparison to all framework? Link to the TF is fine.

2. What do you mean by high level approach like XLA?

**Comment by tkonolige** [comment]
1. I was just following guidelines that I saw here: https://github.com/apache/tvm/pull/6794#issuecomment-719094251.
2. Sorry, not XLA, MLIR. They are taking a TACO approach to sparse matrices https://llvm.discourse.group/t/mlir-support-for-sparse-tensors/2020

**Comment by tqchen** [comment]
Thanks for the discussion so far. In this particular case I agree that having an API review is reasonable(check the possible references in different frameworks). Although it would be great to keep the discussion to this PR and particular operator, so that we can move on constructively :)

**Comment by zhiics** [comment]
@tkonolige Thanks for suggestions. I agree that it might be worth to discuss if a `sparse` namespace is needed although there are already some sparse operators in the code base without using the new namespace.

However, I personally don't think we need to over complicate the process of implementing operators (e.g. comparing across all frameworks) as well, which is also not the process we have taken to implement other operators.

In addition, it may be impractical and/or inefficient to use existing operators to implement some TF/PT ops IMHO. Particularly, there are not that many sparse op in relay to compose a synthetic TF op. 

@codeislife99 could you summarize the discussion so that we can have agreements on all points to move forward?

File: src/relay/op/tensor/transform.cc (hunk 5 above)
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done.

File: python/tvm/relay/op/transform.py (hunk 7 above)
Comment (codeislife99) [reply python/tvm/relay/op/transform.py:1327]: I will include a link to the TF documentation which will make it clearer.

**Comment by codeislife99** [comment]
Context of these PRs: The goal of adding these sparse ops is to enable a customer to run their recommendation model which is currently getting split into multiple subgraphs because of our non-coverage of this op. 

I had an offline discussion with the main reviewers but I will also try to summarize the conclusions from it and the comments here:
1. New namespace : Further discussion will be had in a separate thread, after the current (this one included) sparse ops PRs are merged, since there are a few already existing sparse ops without the namespace and if a new namespace is necessary all the current + previous sparse ops will be put in it. 
2. Documentation : More documentation has been added 
3. High level Approach like XLA: Discussion in a separate thread 
4. This operation is TF specific.

**Comment by codeislife99** [comment]
@tkonolige @mbrookhart Can I get a re-review on this PR ? I have added the TF Frontend code and some more documentation.

**Comment by masahi** [comment]
I'm interested in this discussion. I'm looking at how to support PyTorch [EmbeddingBag](https://pytorch.org/docs/stable/generated/torch.nn.EmbeddingBag.html) op, which is used in Facebook DLRM model and which seems very similar to TF sparse segment sum op this PR is adding. The same ops exist in caffe2, too https://caffe2.ai/docs/sparse-operations.html

So I'd say this op is not specific to TF and generally useful for recsys or possibly NLP use cases. 

This op can be implemented by composing embedding op + reduce, and that is how ONNX exports PyTorch `EmbeddingBag` op. But that would involve materializing the embedding which is huge for DLRM. 

We cannot rely on automatic fusion of embedding + reduce, because our op fusion doesn't fuse any ops before a reduction op (if I remember correctly) cc @jwfromm @mbrookhart

**Comment by codeislife99** [comment]
This functionality and the discussion has been addressed in #7562 . Closing this.
